[windows] Process event data stream

derekwbrown opened the pull request:
Initial revision, starting to get PEDS working on windows.

### What does this PR do?

On windows, implements the network consumer for the process event data stream.
Using the network consumer, gets the APM tags (service/version/env) from the process' directory if configured by `web.config` or `datadog.json` so that the tags will match the APM tags.

### Motivation

Make tagging consistent between APM & USM

### Additional Notes


### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes

To validate that tagging is happening properly.

- install iis
- install agent
- install a properly formatted `datadog.json` or `app.config` into ` C:\windows\System32\WindowsPowerShell\v1.0`.  This allows powershell commands to be tagged.
- stop process agent
  - this is so that you can be sure you're getting all the connections.  At this point, you have 20min to complete the test, which should be no problem
 - Execute a web request against your IIS using powershell, such as: `(start-process powershell -passthru -ArgumentList "-C iwr -usebasicparsing -disablekeepalive -outfile test.txt http://localhost")`
 - Collect the connections as process agent would:
   - `PS C:\windows\System32\WindowsPowerShell\v1.0> $res = ((iwr -UseBasicParsing -DisableKeepAlive http://localhost:3333/net
work_tracer/connections).content | ConvertFrom-Json)`
  - note the assignment to a variable; this is important because you're going to post-process them to validate the results.
  
- the tags are stored separately.  Get a list of tags included in this payload:  
```PS C:\users\Administrator> $res | select -ExpandProperty tags
service:powershell
env:staging
version:somever
http.iis.app_pool:DefaultAppPool
http.iis.sitename:Default Web Site
service:service1
version:1.0-prerelease
http.iis.site:1```

Somewhere should be the tags that you set in the `datadog.json` or `web.config`.  Note that this is an ordered list.

Now, get the connections.  Using this command, you can filter just requests that the remote port was 80, which will hopefully be only one:

`$res.conns | where-object {$_.laddr.port -eq "80" }`
Note the `tags` property should have indices into the list of tags you printed above, and should contain the expected string.


Note:  a subsequent PR will contain some automated testing for this feature.

derekwbrown committed:
[windows] Process event data stream
DO NOT MERGE

Initial revision, starting to get PEDS working on windows.

derekwbrown committed:
undo process consumer; that was the wrong consumer.

Get the network consumer properly initialized, and process events
accordingly.

derekwbrown committed:
remove unneeded file

derekwbrown committed:
Initial revision; using process event data stream, retrieve APM tags when configured
with local config file

derekwbrown committed:
Seems to be working

The driver timestamps the _flow_ with the number of microseconds since system boot.
Convert that to unix time.

hmahmood commented on pkg/network/events/monitor_windows.go:
```diff
@@ -0,0 +1,79 @@
+
+		appcfg, err := iisconfig.ReadDatadogJSON(ddJSON)
+		if err == nil {
+			if err == nil {
```
Extra `err == nil`

derekwbrown replied:
lol. I'm amazed the linter didn't barf all over that.  but thanks, & fixed.

hmahmood commented on pkg/network/events/monitor_windows.go:
```diff
@@ -0,0 +1,79 @@
+	if len(apmtags.DDVersion) > 0 {
+		tags = append(tags, intern.GetByString("version"+":"+apmtags.DDVersion))
+	}
+	if len(tags) == 0 {
```
nit: you can just do `return tags`

derekwbrown replied:
clarified offline and will be leaving as-is.

hmahmood commented on pkg/network/time_converter_windows.go:
```diff
@@ -0,0 +1,42 @@
+
+	// convert this to a unix timestamp
+
+	bootTime := time.Now().Add(-time.Duration(ticks) * time.Millisecond)
```
You can probably cache `bootTime` to avoid a call to `getTickCount64` in the future.

derekwbrown replied:
done.

hmahmood commented on pkg/network/tracer/tracer_windows.go:
```diff
@@ -114,6 +122,23 @@ func NewTracer(config *config.Config, telemetry telemetry.Component) (*Tracer, e
+		if tr.processCache, err = newProcessCache(config.MaxProcessesTracked); err != nil {
+			return nil, fmt.Errorf("could not create process cache; %w", err)
+		}
+		//telemetry.GetCompatComponent().RegisterCollector(tr.processCache)
```
Don't need telemetry?

derekwbrown replied:
added.

hmahmood commented on pkg/network/tracer/tracer_windows.go:
```diff
@@ -114,6 +122,23 @@ func NewTracer(config *config.Config, telemetry telemetry.Component) (*Tracer, e
+		}
+		//telemetry.GetCompatComponent().RegisterCollector(tr.processCache)
+
+		//if tr.timeResolver, err = timeresolver.NewResolver(); err != nil {
+		//	return nil, fmt.Errorf("could not create time resolver: %w", err)
+		//}
```
Probably should get rid of this

derekwbrown replied:
will do.

hmahmood commented on pkg/network/events/monitor.go:
```diff
@@ -136,10 +130,17 @@ func (h *eventConsumerWrapper) Copy(ev *model.Event) any {
 			}
 		}
 	}
+	if !tagsFoundInEnvironment {
```
nit: you can instead just check if `len(p.Tags) == 0` here.

derekwbrown replied:
I could, but I wanted to explicitly call it out.
1) environment variables supercede the various file parsing; my intent was to make the precedence more obvious
2) if someone adds some other way to get the tags, it becomes more obvious that the check for apmtags is related specifically to the environment block.

WDYT?

hmahmood replied:
Sounds good

derekwbrown committed:
linter

richl-dd commented on pkg/network/events/monitor_linux.go:
```diff
@@ -0,0 +1,29 @@
+	if ev.GetEventType() == model.ForkEventType {
+		return ev.GetProcessForkTime()
+	}
+	return time.Time{}
```
I recognize that this block is moved from above into a dedicated function (which is good), so if we don't want to modify then that'd make sense.

I'm curious what the emergent behavior is as the `time` for this process is processed in the agent / backend if we return `time.Time{}` - what do you think about returning an error and skipping this instead? Mostly asking because if we have an EventType that we don't expect, that might be indicative of another problem and this `time.Time{}` return hides.

derekwbrown replied:
So, the reason I broke it out is because linux supports two event types, fork & exec.  windows only supports exec.

The main thing this time is used for (on the windows side) is to correlate processes in the process cache.
Beyond that, objective was to preserve existing behavior.

derekwbrown replied:
to say another way, my objective was to not change the behavior of the pre-existing code.  This maintains the previous behavior.

richl-dd commented on pkg/network/events/monitor_windows.go:
```diff
@@ -0,0 +1,79 @@
+		return ev.GetProcessExecTime()
+	}
+	// windows does not have a fork event.
+	return time.Time{}
```
Similar to above comment in the Linux variant of this function - is it fine to return `time.Time{}` or should we be logging a warning / error and return it that indicates we got an `EventType` we weren't expecting?

richl-dd commented on pkg/network/events/monitor_windows.go:
```diff
@@ -0,0 +1,79 @@
+		return nil
+	}
+
+	// see if there's an app.config in the directory
+	appConfig := filepath.Join(dir, "app.config")
+	ddJSON := filepath.Join(dir, "datadog.json")
+	if _, err := os.Stat(appConfig); err == nil {
+
+		appcfg, err := iisconfig.ReadDotNetConfig(appConfig)
+		if err == nil {
+			found := makeTagsSlice(appcfg)
+			if len(found) > 0 {
+				return found
+			}
+		}
+	}
+	// see if there's a datadog.json
+	if _, err := os.Stat(ddJSON); err == nil {
+
+		appcfg, err := iisconfig.ReadDatadogJSON(ddJSON)
+		if err == nil {
+			if err == nil {
+				found := makeTagsSlice(appcfg)
+				if len(found) > 0 {
+					return found
+				}
+			}
+		}
+	}
```
Couple questions for this block:

1. Is it intentional that the app config content is given precedence over the datadog.json content? Is that documented or did you just happen to choose to parse the app config first? If the order matters, we should confirm that this is the right hierarchy (maybe we already did).
2. For these files, are all environment variables expected to be in each file or could they be split? Right now, if they were split between the files, we'd ignore anything in datadog.json if any (but not all) were defined in the app config
3. Related to (2.) - if there is an app config and at least one environment variable in it, we don't look for datadog.json. In my opinion, we should always check for and parse a datadog.json if it is there, unless we've established and documented to customer that they can only use one or the other AND all environments must be in which ever file they pick.

If we haven't made that clear, we should. Right now, I think for this we should figure out which file has precedence and parse both, with a tie going to whichever we decide has precedence and document it. I suspect since this is new, it isn't documented, so let's decide now and stick with it.

derekwbrown replied:
1. Yes, it's intentional.  Yes it's documented (in the internal doc)
2. I'm not sure what the expected behavior would be in that case, but it seems pathological.  We can check with APM to see how to handle
3. there are no "environment variables" in app config.  app config is allowed to have one of the settings which is essentially a "redirect" to a `datadog.json` file.

richl-dd replied:
1. All good
2. All good
3. What do you mean by there aren't any environment variables? in the `ReadDotNetConfig` function (which I guess is actually reading the `AppConfig`) we look for `DD_` entries, aren't those the "environment variables" that customers define that would normally be actually environment variables on the process?

derekwbrown replied:
those are the key names that happen to be the same as the environment variables.

richl-dd replied:
Thanks for explaining that a process's environment variables don't live in JSON files.

What I meant were the keys in the JSON files that map to environment variables, but I guess what was confusing.

So, question / comment (3.) rewritten to be clearer to you:

Related to (2.) - if there is an app config and at least one ~~environment variable~~ key, we don't look for datadog.json. In my opinion, we should always check for and parse a datadog.json if it is there, unless we've established and documented to customer that they can only use one or the other AND all ~~environments~~ must be in which ever file they pick.

Hopefully that makes more sense.

derekwbrown replied:
I have checked with the APM team and you're right.  I'll update the doc to cover that case, and the code.

richl-dd commented on pkg/network/time_converter_windows.go:
```diff
@@ -0,0 +1,42 @@
+)
+
+var (
+	modkernel          = windows.NewLazyDLL("kernel32.dll")
```
`#include <namingrequest.h>` but could you make this modKernel32 so that we establish a consistent pattern to refer to lazy loaded dll modules?

derekwbrown replied:
done.

richl-dd commented on pkg/network/tracer/tracer_windows.go:
```diff
@@ -59,6 +62,11 @@ type Tracer struct {
+
+	processCache *processCache
+
+	//timeResolver *timeresolver.Resolver
```
Is this commented out by mistake?

derekwbrown replied:
no but should be removed.

richl-dd commented on pkg/network/tracer/tracer_windows.go:
```diff
@@ -301,3 +331,28 @@ func newUSMMonitor(c *config.Config, dh driver.Handle) usm.Monitor {
+		copy(c.Tags, p.Tags)
+	}
+
+	if p.ContainerID != nil {
```
Have we tested this when Windows containers are running on the host? Afaik, we don't support containers at all for USM, regardless of if they're running on the host where the driver is installed, but I don't know if the driver or Agent makes any attempt to distinguish between processes running in a container or directly on the host, so this could give unexpected results if we happen to be running in the presence of containers.

So, I 'm not really sure what the best path forward is here - what do you think?

derekwbrown replied:
I have not (yet) tried in the presence of containers.

shanelhuang replied:
FYI one of the Windows USM pre-reqs is: "Your service is running on a virtual machine."

derekwbrown committed:
remove test that's no longer valid

derekwbrown committed:
review feedback

derekwbrown committed:
add telemetry, and remove dead block

gjulianm reviewed (approved):
LGTM for ebpf-platform files

alexn-dd commented on pkg/network/events/monitor_windows.go:
```diff
@@ -0,0 +1,79 @@
+	"github.com/DataDog/datadog-agent/pkg/util/winutil/iisconfig"
+)
+
+func getProcessStartTime(ev *model.Event) time.Time {
```
Should this default to getting the Win32 `GetProcessTimes`?

derekwbrown replied:
no.  this is querying the process start time stored in the `*model.Event`.  That value is stored at process creation time; and the value needs to be consistent with other areas that read from the `*model.Event`

alexn-dd commented on pkg/network/events/monitor_windows.go:
```diff
@@ -0,0 +1,79 @@
+	}
+	return tags
+}
+func getAPMTags(filename string) []*intern.Value {
```
Nit: need empty line above

derekwbrown replied:
fixed

alexn-dd commented on pkg/network/events/monitor_windows.go:
```diff
@@ -0,0 +1,79 @@
+}
+func getAPMTags(filename string) []*intern.Value {
+
+	dir := filepath.Dir(filename)
```
`filepath.Dir` behavior is atypical.  It seems to default return `.`

alexn-dd replied:
The offical Go API does not define well the error behavior of this function.  Not much that can be done.

alexn-dd commented on pkg/network/events/monitor_windows.go:
```diff
@@ -0,0 +1,79 @@
+	// see if there's an app.config in the directory
+	appConfig := filepath.Join(dir, "app.config")
+	ddJSON := filepath.Join(dir, "datadog.json")
+	if _, err := os.Stat(appConfig); err == nil {
```
Should the error be log somewhere?
In case the file is missing, restricted, corrupted, locked, etc?

derekwbrown replied:
The file is expected to be missing; in fact that's going to be the case in the vast majority of the times this code is executed.

This code path is also expected to be executed a _lot_ (on every process start), so i felt it was better to just continue processing in the case where the file exists (stat returns without error) rather than trying to decode specific errors with a lot of extra logic.

alexn-dd replied:
Can we ignore only the case when the file is missing and not other errors?
How do engineers troubleshoot file related problems in general?

alexn-dd replied:
Offline discussion, this is a hot path and maybe too noisy.

alexn-dd commented on pkg/network/time_converter_windows.go:
```diff
@@ -0,0 +1,42 @@
+
+func driverTimeToUnixTime(t uint64) uint64 {
+	ticks := getTickCount64()
+	// GetTickCount64() returns the number of milliseconds that have elapsed since the system was started
```
Duped comment.

derekwbrown replied:
removed

alexn-dd commented on pkg/network/tracer/tracer_windows.go:
```diff
@@ -59,6 +62,11 @@ type Tracer struct {
+
+	processCache *processCache
+
+	//timeResolver *timeresolver.Resolver
```
Dead code

derekwbrown replied:
fixed

derekwbrown commented on pkg/network/tracer/process_cache.go:
```diff
@@ -196,7 +195,7 @@ func (pc *processCache) Get(pid uint32, ts int64) (*events.Process, bool) {
 	defer pc.mu.Unlock()
 
-	log.TraceFunc(func() string { return fmt.Sprintf("looking up pid %d", pid) })
+	log.TraceFunc(func() string { return fmt.Sprintf("looking up pid %d %v", pid, ts) })
```
%v lets the formatter determine the type.

alexn-dd commented on pkg/network/tracer/tracer_windows.go:
```diff
@@ -301,3 +327,28 @@ func newUSMMonitor(c *config.Config, dh driver.Handle) usm.Monitor {
+
+	c.ContainerID.Source, c.ContainerID.Dest = nil, nil
+
+	// on windows, cLastUpdateEpoch is already set as
```
Nit: `LastUpdateEpoch`

derekwbrown replied:
the member variable name is cLastUpdateEpoch.  Could you clarify what you're expecting here?

alexn-dd replied:
It looked like a typo, but this is fine.

derekwbrown committed:
Redo tag parsing logic based on review feedback.

Prior behavior was stopping when any tag was found, which was incorrect.
continue down precedence list until all tags are found, or precedence list is exhausted.

derekwbrown committed:
review feedback

hmahmood commented on pkg/network/events/monitor.go:
```diff
@@ -136,10 +142,17 @@ func (h *eventConsumerWrapper) Copy(ev *model.Event) any {
+	if len(tagsFound) < len(envFilter) {
+		apmTags := getAPMTags(tagsFound, ev.GetExecFilePath())
+		if len(apmTags) > 0 {
+			p.Tags = append(p.Tags, apmTags...)
```
Can we get dups here?

hmahmood replied:
Oh I see you check that in `getAPMTags`. nvm

derekwbrown committed:
add log message for unexpected error

derekwbrown committed:
add error checking code

derekwbrown committed:
fix initialization during tests

derekwbrown committed:
remove unused definition for macos

derekwbrown commented:
/merge
